CoseIndirectSignature/CoseHashV.cs

```diff
@@ -0,0 +1,330 @@
+    public Task<bool> ContentMatchesAsync(ReadOnlyMemory<byte> data)
+        => Task.FromResult(HashMatches(data: data.ToArray(), stream: null));
+
+    /// <summary>
+    /// Validates that the given data in bytes matches the hash value stored in this instance.
+    /// </summary>
+    /// <param name="data">The data bytes to check to match the hash.</param>
+    /// <returns>True if the hash of <paramref name="data"/> matches HashBytes, False otherwise.</returns>
+    /// <exception cref="ArgumentNullException">Thrown if data passed in is null or has a length of 0.</exception>
+    /// <exception cref="CoseSign1Exception">Thrown if the computed hash length and the stored hash length differ.</exception>
+    public bool ContentMatches(byte[] data)
+        => HashMatches(data: data, stream: null);
+
+    /// <summary>
+    /// Method for handling byte[] and stream for the same logic.
+    /// </summary>
+    /// <param name="data">if specified, then will compute a hash of this data and compare to internal hash value.</param>
+    /// <param name="stream">if data is null and stream is specified, then will compute a hash of this stream and compare to internal hash value.</param>
+    /// <returns>True if the hashes match, False otherwise.</returns>
+    /// <exception cref="ArgumentNullException">Thrown if data is null or data length is 0 and stream is null, or if data is null and stream is null.</exception>
+    /// <exception cref="CoseSign1Exception">Thrown if the length of the computed hash does not match the internal stored hash length, thus the wrong hash algorithm is being used.</exception>
+    private bool HashMatches(byte[]? data, Stream? stream)
+    {
+        // handle input validation
+        if (
+            (data == null || data.Length == 0) &&
+            (stream == null))
+        {
+            throw new ArgumentNullException(nameof(data));
+        }
+
+        // initialize and compute the hash
+        using HashAlgorithm hashAlgorithm = GetHashAlgorithmFromCoseHashAlgorithm(Algorithm);
+        byte[] hash = stream != null ? hashAlgorithm.ComputeHash(stream) : hashAlgorithm.ComputeHash(data);
+
+        // handle the case where the algorithm we derived did not match the algorithm that was used to populate the CoseHashV instance.
+        return hash.Length != HashValue.Length
+            ? throw new CoseSign1Exception($@"The computed hash length of {hash.Length} for hash type {hashAlgorithm.GetType().FullName} created a hash different than the length of {HashValue.Length} which is unexpected.")
+            : hash.SequenceEqual(HashValue);
+    }
+
+    /// <summary>
+    /// Writes the current CoseHashV instance to a cbor byte[].
+    /// </summary>
+    /// <returns>a byte[] cbor representation of the CoseHashV object.</returns>
+    public byte[] Serialize()
+    {
+        CborWriter writer = new(CborConformanceMode.Strict, allowMultipleRootLevelValues: true);
```

Comment:
note that I used `allowMultipleRootLevelValues` here so that I didn't have to manage the properties as an array.  We should discuss on if this is how we want this to be implemented or if we want `allowMultipleRootLevelValues` to be the default of false, and to manage the properties as an array.

Reply:
Interesting, looking at RFC 9054 now I realize that it doesn't specify this explicitly. I would recommend encoding it as an array, for a couple reasons:

- I feel like the +cose-hash-v suffix implies that the content is 1 "thing", rather than an ordered list of things
- In the future, we may want to support the optional location/additional data fields. Encoding the structure as an array makes it easy to see if those are present.

Reply:
I'll push an update to move to that and we'll see how it looks in comparison.

Reply:
latest update moves to array based encoding.